packages/block-editor/src/components/use-block-drop-zone/index.js

```diff
@@ -100,73 +228,71 @@ export default function useBlockDropZone( { element, rootClientId } ) {
 			const blocks = pasteHandler( { HTML, mode: 'BLOCKS' } );
 
 			if ( blocks.length ) {
-				insertBlocks( blocks, blockIndex, rootClientId );
+				insertBlocks( blocks, targetBlockIndex, targetRootClientId );
 			}
 		},
-		[ insertBlocks, blockIndex, rootClientId ]
+		[ insertBlocks, targetBlockIndex, targetRootClientId ]
 	);
 
 	const onDrop = useCallback(
 		( event ) => {
 			const {
-				srcRootClientId,
-				srcClientId,
-				srcIndex,
-				type,
+				srcRootClientId: sourceRootClientId,
+				srcClientId: sourceClientId,
+				srcIndex: sourceBlockIndex,
+				type: dropType,
 			} = parseDropEvent( event );
 
-			const isBlockDropType = ( dropType ) => dropType === 'block';
-			const isSameLevel = ( srcRoot, dstRoot ) => {
-				// Note that rootClientId of top-level blocks will be undefined OR a void string,
-				// so we also need to account for that case separately.
-				return (
-					srcRoot === dstRoot ||
-					( ! srcRoot === true && ! dstRoot === true )
-				);
-			};
-			const isSameBlock = ( src, dst ) => src === dst;
-			const isSrcBlockAnAncestorOfDstBlock = ( src, dst ) =>
-				getClientIdsOfDescendants( [ src ] ).some(
-					( id ) => id === dst
-				);
+			// If the user isn't dropping a block, return early.
+			if ( dropType !== 'block' ) {
+				return;
+			}
+
+			// If the user is dropping to the same position, return early.
+			if (
+				sourceRootClientId === targetRootClientId &&
+				sourceBlockIndex === targetBlockIndex
+			) {
+				return;
+			}
 
+			// If the user is attempting to drop a block within its own
+			// nested blocks, return early as this would create infinite
+			// recursion.
 			if (
-				! isBlockDropType( type ) ||
-				isSameBlock( srcClientId, clientId ) ||
-				isSrcBlockAnAncestorOfDstBlock(
-					srcClientId,
-					clientId || rootClientId
+				targetRootClientId === sourceClientId ||
+				getClientIdsOfDescendants( [ sourceClientId ] ).some(
+					( id ) => id === targetRootClientId
 				)
 			) {
 				return;
 			}
 
-			const dstIndex = clientId
-				? getBlockIndex( clientId, rootClientId )
-				: undefined;
-			const positionIndex = blockIndex;
+			const isAtSameLevel =
+				sourceRootClientId === targetRootClientId ||
+				( sourceRootClientId === '' &&
+					targetRootClientId === undefined );
```

Comment:
@talldan: Could we converge on the types of these two expectably similar variables? If that's not possible, can we add comments explaining the discrepancy?

Reply:
@mcsf I've put together this PR that improves things - https://github.com/WordPress/gutenberg/pull/24307